File: java/45.java
class Solution {
    public int jump(int[] nums) {
        int n = nums.length;
        int end = 0, farthest = 0, steps = 0;

        for (int i=0;i<n;i++) {
            farthest = Math.max(farthest, nums[i] + i);
            if (i == end) {
                end = farthest;
                if (i != n - 1)
                    steps++;
            }
        }

        return steps;
    }
}

File: py/45.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return 0
            
        cur = 0
        left = nums[0]
        farthest = nums[0]
        farthestIdx = 0
        ans = 1
        while cur + 1 < len(nums):
            
            if left == 0:
                # print(cur, left, farthest)
                ans += 1
                left = farthest - cur
                
            cur += 1
            left -= 1
            if nums[cur] + cur > farthest:
                farthest = nums[cur] + cur
                farthestIdx = cur

        return ans
                
